packages/react-native/src/storage/Chunkifier.ts

```diff
@@ -0,0 +1,120 @@
+
+export interface ChunkifierOptions<W extends Chunk> {
+    /**
+     * Chunk splitter factory. The factory will be called when creating a new chunk.
+     */
+    readonly splitter: ChunkSplitterFactory<W>;
+
+    /**
+     * Chunk sink. The sink will be called when creating a new chunk.
+     */
+    readonly sink: ChunkSink<W>;
+
+    readonly allowEmptyChunks?: boolean;
+}
+
+interface StreamContext<W extends Chunk> {
+    readonly stream: WritableStream<W>;
+    readonly streamWriter: WritableStreamDefaultWriter<W>;
+    isEmptyChunk: boolean;
+}
+
+export class ChunkifierSink<W extends Chunk> implements UnderlyingSink<W> {
+    private _context?: StreamContext<W>;
+    private _splitter?: ChunkSplitter<W>;
+    private _chunkCount = 0;
+
+    constructor(private readonly _options: ChunkifierOptions<W>) {}
+
+    public async write(data: W): Promise<void> {
+        // If data is empty from the start, forward the write directly to current stream
+        if (this.isEmpty(data)) {
+            return await this.ensureStreamContext().streamWriter.write(data);
+        }
+
+        while (data) {
+            if (this.isEmpty(data)) {
+                break;
+            }
+
+            const splitter = this.ensureSplitter();
+            const [currentChunk, nextChunk] = splitter(data);
+            if (nextChunk === undefined) {
+                const current = this.ensureStreamContext();
+                if (!this.isEmpty(currentChunk)) {
+                    current.isEmptyChunk = false;
+                }
+
+                return await current.streamWriter.write(currentChunk);
+            }
+
+            data = nextChunk;
+            if (
+                this._context
+                    ? this._context.isEmptyChunk
+                    : this.isEmpty(currentChunk) && !this._options.allowEmptyChunks
+            ) {
+                continue;
+            }
+
+            const current = this.ensureStreamContext();
+            await current.streamWriter.write(currentChunk);
+            current.streamWriter.releaseLock();
+
+            // On next loop iteration, or write, create new stream again
+            this.reset();
+        }
+    }
+
+    public async close() {
+        return await this._context?.streamWriter.close();
+    }
+
+    private ensureStreamContext() {
+        if (!this._context) {
+            return (this._context = this.createStreamContext());
+        }
+        return this._context;
+    }
+
+    private ensureSplitter() {
+        if (!this._splitter) {
+            return (this._splitter = this._options.splitter());
+        }
+        return this._splitter;
+    }
+
+    private createStreamContext(): StreamContext<W> {
+        const stream = this._options.sink(this._chunkCount++);
+        const writer = stream.getWriter();
+        return { stream, streamWriter: writer, isEmptyChunk: true };
+    }
+
+    private reset() {
+        this._context = undefined;
+        this._splitter = undefined;
```

Comment:
problem:

I needed to check multiple times splitter definition and implementation to understand every time when you reset the splitter you're changing the current line. Is it worth explaining it somewhere here?

Reply:
I added some comments.